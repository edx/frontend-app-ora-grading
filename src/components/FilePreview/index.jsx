export { default as FileCard } from './FileCard';
export { default as ImageRenderer } from './ImageRenderer';
export { default as PDFRenderer } from './PDFRenderer';
