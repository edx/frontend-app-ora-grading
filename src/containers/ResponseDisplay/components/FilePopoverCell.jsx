import React from 'react';
import PropTypes from 'prop-types';

import InfoPopover from 'components/InfoPopover';

export const FilePopoverCell = ({ row: { original } }) => (
  <InfoPopover>
    <div className="help-popover-option">
      <strong>File Name</strong>
      <br />
      {original.name}
    </div>
    <div className="help-popover-option">
      <strong>Description</strong>
      <br />
      {original.description}
    </div>
  </InfoPopover>
);

FilePopoverCell.defaultProps = {
  row: {
    original: {},
  },
};

FilePopoverCell.propTypes = {
  row: PropTypes.shape({
    original: PropTypes.shape({
      name: PropTypes.string.isRequired,
      description: PropTypes.string,
      downloadUrl: PropTypes.string,
    }),
  }),
};

export default FilePopoverCell;
