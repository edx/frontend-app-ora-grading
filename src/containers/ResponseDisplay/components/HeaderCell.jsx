import React from 'react';

import InfoPopover from 'components/InfoPopover';

/* eslint react/prop-types: 0 */
export const HeaderEllipsesCell = ({ value }) => (
  <div className="text-truncate">{value}</div>
);

/* eslint react/prop-types: 0 */
export const HeaderFileExtensionCell = ({ value = '' }) => {
  const extension = value.split('.')?.pop().toUpperCase();
  return <div className="text-truncate">{extension}</div>;
};

/* eslint react/prop-types: 0 */
export const HeaderInfoPopoverCell = ({ row }) => {
  const { original } = row;
  return (
    <InfoPopover>
      {Object.keys(original).map((key) => (
        <div key={key} className="help-popover-option">
          <strong>{key.toUpperCase()}</strong>
          <br />
          {original[key]}
        </div>
      ))}
    </InfoPopover>
  );
};

export default {
  HeaderEllipsesCell,
  HeaderFileExtensionCell,
  HeaderInfoPopoverCell,
};
