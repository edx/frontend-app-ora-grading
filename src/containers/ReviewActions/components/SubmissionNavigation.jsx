import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import { Icon, IconButton } from '@edx/paragon';
import { ChevronLeft, ChevronRight } from '@edx/paragon/icons';
import { FormattedMessage, injectIntl, intlShape } from '@edx/frontend-platform/i18n';

import selectors from 'data/selectors';
import thunkActions from 'data/thunkActions';
import messages from './messages';

/**
 * <SubmissionNavigation />
 */
export const SubmissionNavigation = ({
  intl,
  hasPrevSubmission,
  hasNextSubmission,
  loadPrev,
  loadNext,
  activeIndex,
  selectionLength,
}) => (
  <span className="submission-navigation">
    <IconButton
      className="ml-1"
      size="inline"
      disabled={!hasPrevSubmission}
      alt={intl.formatMessage(messages.loadPrevious)}
      src={ChevronLeft}
      iconAs={Icon}
      onClick={loadPrev}
    />
    <span className="ml-1">
      <FormattedMessage
        {...messages.navigationLabel}
        values={{ current: activeIndex + 1, total: selectionLength }}
      />
    </span>
    <IconButton
      className="ml-1"
      size="inline"
      disabled={!hasNextSubmission}
      alt={intl.formatMessage(messages.loadNext)}
      src={ChevronRight}
      iconAs={Icon}
      onClick={loadNext}
    />
  </span>
);
SubmissionNavigation.defaultProps = {
  hasPrevSubmission: false,
  hasNextSubmission: false,
};
SubmissionNavigation.propTypes = {
  // injected
  intl: intlShape.isRequired,
  // redux
  activeIndex: PropTypes.number.isRequired,
  hasNextSubmission: PropTypes.bool,
  hasPrevSubmission: PropTypes.bool,
  loadNext: PropTypes.func.isRequired,
  loadPrev: PropTypes.func.isRequired,
  selectionLength: PropTypes.number.isRequired,
};

export const mapStateToProps = (state) => ({
  activeIndex: selectors.grading.activeIndex(state),
  hasNextSubmission: selectors.grading.next.doesExist(state),
  hasPrevSubmission: selectors.grading.prev.doesExist(state),
  selectionLength: selectors.grading.selectionLength(state),
});

export const mapDispatchToProps = {
  loadNext: thunkActions.grading.loadNext,
  loadPrev: thunkActions.grading.loadPrev,
};

export default injectIntl(connect(mapStateToProps, mapDispatchToProps)(SubmissionNavigation));
