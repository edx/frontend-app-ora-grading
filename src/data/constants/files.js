import { StrictDict } from 'utils';

export const FileTypes = StrictDict({
  pdf: 'pdf',
  jpg: 'jpg',
  jpeg: 'jpeg',
  png: 'png',
  bmp: 'bmp',
});

export default FileTypes;
