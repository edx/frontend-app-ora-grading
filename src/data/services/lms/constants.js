import { StrictDict } from 'utils';

export const lockStatuses = StrictDict({
  unlocked: 'unlocked',
  locked: 'locked',
  inProgress: 'in-progress',
});

export const gradeStatuses = StrictDict({
  ungraded: 'ungraded',
  graded: 'graded',
});

export const gradingStatuses = StrictDict({
  ungraded: gradeStatuses.ungraded,
  graded: gradeStatuses.graded,
  locked: lockStatuses.locked,
  inProgress: lockStatuses.inProgress,
});

export const gradingStatusDisplay = StrictDict({
  [gradingStatuses.ungraded]: 'Ungraded',
  [gradingStatuses.locked]: 'Currently being graded by someone else',
  [gradingStatuses.graded]: 'Grading Complete',
  [gradingStatuses.inProgress]: 'You are currently grading this response',
});

export const feedbackRequirement = StrictDict({
  disabled: 'disabled',
  required: 'required',
  optional: 'optional',
});
