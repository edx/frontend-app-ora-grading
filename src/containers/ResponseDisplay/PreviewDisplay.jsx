import React from 'react';
import PropTypes from 'prop-types';

import { StrictDict } from 'utils';
import { FileTypes } from 'data/constants/files';

import { FileCard, PDFRenderer, ImageRenderer } from 'components/FilePreview';

import './PreviewDisplay.scss';

/**
 * <PreviewDisplay />
 */
export class PreviewDisplay extends React.Component {
  static RENDERERS = StrictDict({
    [FileTypes.pdf]: PDFRenderer,
    [FileTypes.jpg]: ImageRenderer,
    [FileTypes.jpeg]: ImageRenderer,
    [FileTypes.bmp]: ImageRenderer,
    [FileTypes.png]: ImageRenderer,
  });

  static SUPPORTED_TYPES = Object.keys(PreviewDisplay.RENDERERS);

  constructor(props) {
    super(props);
    this.isSupported = this.isSupported.bind(this);
    this.fileType = this.fileType.bind(this);
  }

  get supportedFiles() {
    return this.props.files.filter(this.isSupported);
  }

  isSupported(file) {
    return PreviewDisplay.SUPPORTED_TYPES.includes(this.fileType(file.name));
  }

  fileType(fileName) {
    return fileName.split('.').pop();
  }

  render() {
    return (
      <>
        {this.supportedFiles.map(({ name, downloadUrl }) => {
          const Renderer = PreviewDisplay.RENDERERS[this.fileType(name)];
          return (
            <FileCard key={downloadUrl} name={name}>
              <Renderer fileName={name} url={downloadUrl} />
            </FileCard>
          );
        })}
      </>
    );
  }
}

PreviewDisplay.defaultProps = {
  files: [],
};
PreviewDisplay.propTypes = {
  files: PropTypes.arrayOf(PropTypes.shape({
    name: PropTypes.string,
    downloadUrl: PropTypes.string,
  })),
};

export default PreviewDisplay;
