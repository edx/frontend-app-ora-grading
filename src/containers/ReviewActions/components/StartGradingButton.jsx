import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import { Button } from '@edx/paragon';
import { Cancel, Highlight } from '@edx/paragon/icons';
import { FormattedMessage } from '@edx/frontend-platform/i18n';

import { selectors, thunkActions } from 'data/redux';
import { RequestKeys } from 'data/constants/requests';
import { gradingStatuses as statuses } from 'data/services/lms/constants';

import StopGradingConfirmModal from './StopGradingConfirmModal';
import OverrideGradeConfirmModal from './OverrideGradeConfirmModal';
import messages from './messages';

export const buttonArgs = {
  [statuses.ungraded]: {
    label: messages.startGrading,
    iconAfter: Highlight,
  },
  [statuses.graded]: {
    label: messages.overrideGrade,
    iconAfter: Highlight,
  },
  [statuses.inProgress]: {
    label: messages.stopGrading,
    iconAfter: Cancel,
  },
};

export class StartGradingButton extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      showConfirmStopGrading: false,
      showConfirmOverrideGrade: false,
    };

    this.showConfirmStopGrading = this.showConfirmStopGrading.bind(this);
    this.hideConfirmStopGrading = this.hideConfirmStopGrading.bind(this);
    this.showConfirmOverrideGrade = this.showConfirmOverrideGrade.bind(this);
    this.hideConfirmOverrideGrade = this.hideConfirmOverrideGrade.bind(this);
    this.confirmStopGrading = this.confirmStopGrading.bind(this);
    this.confirmOverrideGrade = this.confirmOverrideGrade.bind(this);
    this.handleClick = this.handleClick.bind(this);
  }

  showConfirmStopGrading() {
    this.setState({ showConfirmStopGrading: true });
  }

  hideConfirmStopGrading() {
    this.setState({ showConfirmStopGrading: false });
  }

  showConfirmOverrideGrade() {
    this.setState({ showConfirmOverrideGrade: true });
  }

  hideConfirmOverrideGrade() {
    this.setState({ showConfirmOverrideGrade: false });
  }

  confirmStopGrading() {
    this.hideConfirmStopGrading();
    this.props.stopGrading();
  }

  confirmOverrideGrade() {
    this.hideConfirmOverrideGrade();
    this.props.startGrading();
  }

  handleClick() {
    if (this.props.gradingStatus === statuses.inProgress) {
      this.showConfirmStopGrading();
    } else if (this.props.gradingStatus === statuses.graded) {
      this.showConfirmOverrideGrade();
    } else {
      this.props.startGrading();
    }
  }

  render() {
    const { gradingStatus } = this.props;
    if (gradingStatus === statuses.locked) {
      return null;
    }
    const args = buttonArgs[gradingStatus];
    return (
      <>
        <Button
          variant="primary"
          iconAfter={args.iconAfter}
          onClick={this.handleClick}
          disabled={this.props.gradeIsPending || this.props.lockIsPending}
        >
          <FormattedMessage {...args.label} />
        </Button>
        <OverrideGradeConfirmModal
          isOpen={this.state.showConfirmOverrideGrade}
          onCancel={this.hideConfirmOverrideGrade}
          onConfirm={this.confirmOverrideGrade}
        />
        <StopGradingConfirmModal
          isOpen={this.state.showConfirmStopGrading}
          onCancel={this.hideConfirmStopGrading}
          onConfirm={this.confirmStopGrading}
          isOverride={this.props.gradeStatus === statuses.graded}
        />
      </>
    );
  }
}

StartGradingButton.propTypes = {
  gradeStatus: PropTypes.string.isRequired,
  gradingStatus: PropTypes.string.isRequired,
  startGrading: PropTypes.func.isRequired,
  stopGrading: PropTypes.func.isRequired,
  gradeIsPending: PropTypes.bool.isRequired,
  lockIsPending: PropTypes.bool.isRequired,
};

export const mapStateToProps = (state) => ({
  gradeIsPending: selectors.requests.isPending(state, { requestKey: RequestKeys.submitGrade }),
  lockIsPending: selectors.requests.isPending(state, { requestKey: RequestKeys.setLock }),
  gradeStatus: selectors.grading.selected.gradeStatus(state),
  gradingStatus: selectors.grading.selected.gradingStatus(state),
});

export const mapDispatchToProps = {
  startGrading: thunkActions.grading.startGrading,
  stopGrading: thunkActions.grading.cancelGrading,
};

export default connect(mapStateToProps, mapDispatchToProps)(StartGradingButton);
