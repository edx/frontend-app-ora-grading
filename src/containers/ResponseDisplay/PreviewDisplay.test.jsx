import React from 'react';
import { shallow } from 'enzyme';

import { FileTypes } from 'data/constants/files';
import { FileRenderer } from 'components/FilePreview';
import { PreviewDisplay } from './PreviewDisplay';

jest.mock('components/FilePreview', () => ({
  FileRenderer: () => 'FileRenderer',
  isSupported: (file) => jest.requireActual('components/FilePreview').isSupported(file),
}));

describe('PreviewDisplay', () => {
  describe('component', () => {
    const supportedTypes = [
      FileTypes.pdf,
      FileTypes.jpg,
      FileTypes.jpeg,
      FileTypes.bmp,
      FileTypes.png,
      FileTypes.txt,
    ];
    const props = {
      files: [
        ...supportedTypes.map((fileType, index) => ({
          name: `fake_file_${index}.${fileType}`,
          description: `file description ${index}`,
          downloadUrl: `/url-path/fake_file_${index}.${fileType}`,
        })),
        {
          name: 'bad_ext_fake_file.other',
          description: 'bad_ext file description',
          downloadUrl: 'bad_ext.other',
        },
      ],
    };
    let el;
    beforeEach(() => {
      el = shallow(<PreviewDisplay {...props} />);
    });

    describe('snapshot', () => {
      test('files does not exist', () => {
        expect(el).toMatchSnapshot();
      });
      test('files exited for props', () => {
        el.setProps({ files: [] });
        expect(el).toMatchSnapshot();
      });
    });

    describe('component', () => {
      test('only renders compatible files', () => {
        const cards = el.find(FileRenderer);
        expect(cards.length).toEqual(supportedTypes.length);
        new Array(cards.length).fill(0).forEach((_, index) => {
          expect(
            cards.at(index).prop('file'),
          ).toEqual(props.files[index]);
        });
      });
    });
  });
});
