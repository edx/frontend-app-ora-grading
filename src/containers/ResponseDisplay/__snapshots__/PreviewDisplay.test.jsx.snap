// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PreviewDisplay component snapshot files does not exist 1`] = `
<Card
  className="submission-files"
>
  <Collapsible.Advanced
    defaultOpen={true}
    key="some file name.jpg"
  >
    <Collapsible.Trigger
      className="submission-files-title"
    >
      <h3>
        some file name.jpg
      </h3>
      <Collapsible.Visible
        whenClosed={true}
      >
        <Icon
          src={[MockFunction icons.ArrowDropDown]}
        />
      </Collapsible.Visible>
      <Collapsible.Visible
        whenOpen={true}
      >
        <Icon
          src={[MockFunction icons.ArrowDropUp]}
        />
      </Collapsible.Visible>
    </Collapsible.Trigger>
    <Collapsible.Body
      className="submission-files-body"
    >
      <PreviewPanel />
    </Collapsible.Body>
  </Collapsible.Advanced>
  <Collapsible.Advanced
    defaultOpen={true}
    key="file number 2.pdf"
  >
    <Collapsible.Trigger
      className="submission-files-title"
    >
      <h3>
        file number 2.pdf
      </h3>
      <Collapsible.Visible
        whenClosed={true}
      >
        <Icon
          src={[MockFunction icons.ArrowDropDown]}
        />
      </Collapsible.Visible>
      <Collapsible.Visible
        whenOpen={true}
      >
        <Icon
          src={[MockFunction icons.ArrowDropUp]}
        />
      </Collapsible.Visible>
    </Collapsible.Trigger>
    <Collapsible.Body
      className="submission-files-body"
    >
      <PreviewPanel />
    </Collapsible.Body>
  </Collapsible.Advanced>
</Card>
`;

exports[`PreviewDisplay component snapshot files exited for props 1`] = `""`;
